compiler/src/dmd/expressionsem.d

```diff
@@ -5115,7 +5115,13 @@ private extern (C++) final class ExpressionSemanticVisitor : Visitor
                 tb = tb.isTypeDArray().next.toBasetype();
             }
 
-            if (global.params.betterC || !sc.needsCodegen())
+            if (global.params.betterC && sc.needsCodegen())
+            {
+                error(exp.loc, "expression `%s` allocates with the GC and cannot be used with switch `-betterC`", exp.toChars());
```

Comment:
Need something better than referencing switches that don't exist in error messages.  #15747

Reply:
Perhaps we need an enum that maps fields in `dmd.globals.Params` to their corresponding CLI option name. That way GDC and LDC can override DMD-specific CLI option names in one place without requiring ugly conditional code like #15747.

Reply:
Yeah, the `version(IN_GCC)` is awful. However, it is not the scope of this PR to replace it.

Reply:
fair enough